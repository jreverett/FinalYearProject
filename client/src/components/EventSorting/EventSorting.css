.sorting-container {
  display: flex;
  padding-left: 10px;
}

#sorting-distance-tooltip.disabled .dropdown {
  pointer-events: none;
}

.button-sorting-selector {
  padding-right: 15px;
}

.button-sorting-selector,
.dropdown-sorting-selector {
  background-color: transparent;
  border: none;
  color: rgb(77, 77, 77);
  font-weight: 600;
}

.button-sorting-selector,
.dropdown-sorting-selector {
  /* overrule bootstrap CSS */
  color: rgb(77, 77, 77) !important;
  background-color: transparent !important;
}

.show > .btn-info.dropdown-sorting-selector {
  background-color: transparent;
  color: black;
}
