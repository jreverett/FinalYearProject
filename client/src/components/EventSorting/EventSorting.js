import React, { Component } from 'react';
import { Button, Dropdown, Tooltip, OverlayTrigger } from 'react-bootstrap';
import { ConditionalWrapper } from '../../components';
import { userService } from '../../services';
import { haversineDistance } from '../../utilities';
import '../../common.css';
import './EventSorting.css';

class EventSorting extends Component {
  constructor(props) {
    super(props);

    this.state = {
      allowDistance: false,
    };
  }

  componentDidMount() {
    this.setState({
      allowDistance: userService.loggedInUserValue?.address,
    });
  }

  renderTooltip() {
    return (
      <Tooltip id="dropdown-tooltip">
        An account with an address is required to use this function
      </Tooltip>
    );
  }

  sortBy = (criteria) => {
    let events = this.props.events.data;
    const updateEvents = this.props.updateEvents;

    if (!events) return;

    switch (criteria) {
      case 'reccomended':
        // TODO - based on mix of other criteria
        break;
      case 'popularity':
        return updateEvents({
          data: events.sort(
            (a, b) => b.subscribers.length - a.subscribers.length
          ),
        });
      case 'distance-close-far':
        var distanceAsc = this.getEventDistances().sort((a, b) => a[1] - b[1]);
        distanceAsc = distanceAsc.map((d) => d[0]);
        return updateEvents({ data: distanceAsc });
      case 'distance-far-close':
        var distanceDesc = this.getEventDistances().sort((a, b) => b[1] - a[1]);
        distanceDesc = distanceDesc.map((d) => d[0]);
        return updateEvents({ data: distanceDesc });
      case 'cost-low-high':
        return updateEvents({ data: events.sort((a, b) => a.cost - b.cost) });
      case 'cost-high-low':
        return updateEvents({ data: events.sort((a, b) => b.cost - a.cost) });
      case 'date-earliest-latest':
        return updateEvents({
          data: events.sort(
            (a, b) => Date.parse(a.start) - Date.parse(b.start)
          ),
        });
      case 'date-latest-earliest':
        return updateEvents({
          data: events.sort(
            (a, b) => Date.parse(b.start) - Date.parse(a.start)
          ),
        });
      default:
        return;
    }
  };

  getEventDistances() {
    const userAddress = this.props.loggedInUser.address;
    var distances = [];
    this.props.events.data.forEach((event) => {
      distances.push([event, haversineDistance(userAddress, event.address)]);
    });

    return distances;
  }

  render() {
    const allowDistance = this.state.allowDistance;
    return (
      <div className="sorting-container">
        {/* RECCOMENDED */}
        <Button
          className="button-sorting-selector"
          onClick={() => this.sortBy('reccomended')}
        >
          Reccomended
        </Button>

        {/* POPULARITY */}
        <Button
          className="button-sorting-selector"
          onClick={() => this.sortBy('popularity')}
        >
          Most Popular
        </Button>

        {/* DISTANCE */}
        <ConditionalWrapper
          condition={!allowDistance}
          wrapper={(children) => (
            <OverlayTrigger placement="right" overlay={this.renderTooltip()}>
              {children}
            </OverlayTrigger>
          )}
        >
          <div
            id="sorting-distance-tooltip"
            className={allowDistance ? '' : ' disabled'}
          >
            <Dropdown onClick={() => this.sortBy('distance')}>
              <Dropdown.Toggle
                className="dropdown-sorting-selector"
                variant="info"
                disabled={!allowDistance}
              >
                Distance
              </Dropdown.Toggle>
              <Dropdown.Menu>
                <Dropdown.Item
                  onClick={() => this.sortBy('distance-close-far')}
                >
                  Closest First
                </Dropdown.Item>
                <Dropdown.Item
                  onClick={() => this.sortBy('distance-far-close')}
                >
                  Farthest First
                </Dropdown.Item>
              </Dropdown.Menu>
            </Dropdown>
          </div>
        </ConditionalWrapper>

        {/* COST */}
        <Dropdown>
          <Dropdown.Toggle className="dropdown-sorting-selector" variant="info">
            Cost
          </Dropdown.Toggle>
          <Dropdown.Menu>
            <Dropdown.Item onClick={() => this.sortBy('cost-low-high')}>
              Low to High
            </Dropdown.Item>
            <Dropdown.Item onClick={() => this.sortBy('cost-high-low')}>
              High to Low
            </Dropdown.Item>
          </Dropdown.Menu>
        </Dropdown>

        {/* START DATE */}
        <Dropdown>
          <Dropdown.Toggle className="dropdown-sorting-selector" variant="info">
            Date
          </Dropdown.Toggle>
          <Dropdown.Menu>
            <Dropdown.Item onClick={() => this.sortBy('date-earliest-latest')}>
              Soonest First
            </Dropdown.Item>
            <Dropdown.Item onClick={() => this.sortBy('date-latest-earliest')}>
              Latest First
            </Dropdown.Item>
          </Dropdown.Menu>
        </Dropdown>
      </div>
    );
  }
}

export default EventSorting;
