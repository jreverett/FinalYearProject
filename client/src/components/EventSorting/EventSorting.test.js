import React from 'react';
import renderer from 'react-test-renderer';
import EventSorting from './EventSorting';

describe('<EventSorting />', () => {
  it('matches the snapshot', () => {
    const tree = renderer.create(<EventSorting />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
