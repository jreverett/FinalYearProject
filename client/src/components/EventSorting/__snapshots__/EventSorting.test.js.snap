// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventSorting /> matches the snapshot 1`] = `
<div
  className="sorting-container"
>
  <button
    className="button-sorting-selector disabled btn btn-primary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    Recomended
  </button>
  <button
    className="button-sorting-selector btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Most Popular
  </button>
  <div
    className=" disabled"
    id="sorting-distance-tooltip"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="dropdown"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      <button
        aria-expanded={false}
        aria-haspopup={true}
        className="dropdown-sorting-selector dropdown-toggle btn btn-info"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Distance
      </button>
    </div>
  </div>
  <div
    className="dropdown"
    onKeyDown={[Function]}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-sorting-selector dropdown-toggle btn btn-info"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Cost
    </button>
  </div>
  <div
    className="dropdown"
    onKeyDown={[Function]}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-sorting-selector dropdown-toggle btn btn-info"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Date
    </button>
  </div>
</div>
`;
